### PR TITLE
Reduce CI time

### LINKED ISSUE #2127: Reduce Windows CI time
The build of Windows CI is almost doubling the time to run CI. It is expected for it to be a longer than other OSes because it includes the .NET Framework in build and tests. That said we should consider if it worth separating, at least the tests, in multiple runners to reduce the CI duration: currently around ~26 minutes, based on the other OSes we could have it ~15 minutes.

### PR DESCRIPTION
**Pull request by pjanotti**
## Why

Reduce the wall clock time for each CI - it does increase the time

Fixes #2127

## What

- Separate build and test workflows, instead of a monolithic workflow, this result in new jobs in the CI workflow. A further breakdown could avoid some redundancy introduced with this change, e.g.: native tests are being built and running multiple times.
- Cache NuGet restore step so it can be re-used between CI runs, this required a new parameter to the build: `--no-restore` similar to one used with dotnet CLI build and publish.
- Separate the tests using Windows docker images to their own job (this became the long poll after initial changes). A follow-up to this would be to cache the docker images to avoid re-downloading them on every run.
- Allow targeting specific TFMs for tests, at the cost of some redundancy, this allowed to parallelize tests for different TFMs (this was also a critical path on Windows builds).

## Tests

N/A

## Checklist

~~- [ ] `CHANGELOG.md` is updated.~~

~~- [ ] Documentation is updated.~~

~~- [ ] New features are covered by tests.~~

### CODE REVIEW

**Comment by pjanotti** [comment]
The first run will be slower because the NuGet packages were not cached yet, so from time to time, when NuGet package caches are not hit there will be some longer runs.

**Comment by pjanotti** [comment]
image

https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/actions/runs/5161224056?pr=2598

**Review by pellared — COMMENTED** [review]
Making it as 4 PRs would make the review easier. Moreover the impact of each PR would be more visible. At last it would be easier to revert a change in case of any problems.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -14,34 +14,41 @@ jobs:
...
+        id: nuget-cache
         with:
-          name: logs-${{ matrix.machine }}
-          path: build_data/
+          key: ${{ hashFiles('**/Directory.packages.props', './build/LibraryVersions.g.cs', '**/packages.config' ) }}
+          path: ${{ env.NUGET_PACKAGES }}
+      - if: ${{ steps.nuget-cache.outputs.cache-hit != 'true' }}
+        name: Restore NuGet packages
+        continue-on-error: false
+        run: ./build.cmd Restore
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply .github/workflows/ci.yml:42]: I did this separately initially on my repo, the problem is that without it making the tests concurrent ends up being too slow: this is one of the most expensive steps on the workflow, the restore itself can take between 4 to 6 minutes, while the action/cache typically is under a minute. The way that it is setup even if the build jobs are a cache miss the test jobs will hit the cache.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -172,6 +196,172 @@ jobs:
...
+            machine: macos-11
+          - test-tfm: net462
+            machine: ubuntu-20.04
+        include:
+          - machine: windows-2022
+            containers: none
+            log-dir: "/c/ProgramData/OpenTelemetry .NET AutoInstrumentation/logs"
+          - machine: ubuntu-20.04
+            containers: linux
+            log-dir: "/var/log/opentelemetry/dotnet"
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply .github/workflows/ci.yml:217]: My understanding is that we need to always checkout the repo. AFAIK even if the runner is re-used one has to assume that the repo is not there yet.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -172,6 +196,172 @@ jobs:
...
+        with:
+          dotnet-version: | 
+            6.0.408
+            7.0.302
+      - name: Check for NuGet packages cache
+        uses: actions/cache@v3
+        id: nuget-cache
+        with:
+          key: ${{ hashFiles('**/Directory.packages.props', './build/LibraryVersions.g.cs', '**/packages.config' ) }}
+          path: ${{ env.NUGET_PACKAGES }}
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply .github/workflows/ci.yml:322]: Yeah, I'm unsatisfied with all this repetition. Will do in a separate PR.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -172,6 +196,172 @@ jobs:
...
+        machine: [ windows-2022, ubuntu-20.04, macos-11 ]
+        exclude:
+          - test-tfm: net462
+            machine: macos-11
+          - test-tfm: net462
+            machine: ubuntu-20.04
+        include:
+          - machine: windows-2022
+            containers: none
+            log-dir: "/c/ProgramData/OpenTelemetry .NET AutoInstrumentation/logs"
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply .github/workflows/ci.yml]: af12f1f7362dc30d683180fe547115108bfd1ab5

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -172,6 +196,172 @@ jobs:
...
+          - machine: ubuntu-20.04
+            containers: linux
+            log-dir: "/var/log/opentelemetry/dotnet"
+          - machine: macos-11
+            containers: none
+            log-dir: "/var/log/opentelemetry/dotnet"
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply .github/workflows/ci.yml]: af12f1f7362dc30d683180fe547115108bfd1ab5

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -14,34 +14,41 @@ jobs:
...
-        uses: potatoqualitee/mssqlsuite@v1.7
-        if: ${{ runner.os == 'Windows' }}
-        with: 
-          install: localdb
-      - run: ./build.cmd Workflow --containers ${{ matrix.containers }}
-      - name: Upload logs
-        uses: actions/upload-artifact@v3.1.2
-        if: always()
+      - name: Check for NuGet packages cache
+        uses: actions/cache@v3
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply .github/workflows/ci.yml]: 6c5a20b77b8ec6979f9e1c16881d8a6e2dfe5b5f

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -14,34 +14,41 @@ jobs:
...
           - machine: ubuntu-20.04
-            containers: linux
             log-dir: "/var/log/opentelemetry/dotnet"
           - machine: macos-11
-            containers: none
             log-dir: "/var/log/opentelemetry/dotnet"
     runs-on: ${{ matrix.machine }}
+    env:
+      NUGET_PACKAGES: ${{ github.workspace }}/packages
+      DOTNET_CLI_TELEMETRY_OPTOUT: 1
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply .github/workflows/ci.yml]: ca6aef1c4881448e19a54e73d5670f957066b192

**Comment by pjanotti** [comment]
> Making it as 4 PRs would make the review easier. Moreover the impact of each PR would be more visible. At last it would be easier to revert a change in case of any problems.

I agree with the sentiment, it could be broke in more PRs, got carried with the fact that I need some of the changes to measure the overall impact, eg.: separating test TFM really needs the NuGet cache. That said I recognize that in the end it got a bit too large with many things in a single PR. Sorry!

File: build/Build.cs [hunk 8]
```diff
@@ -15,13 +15,17 @@ partial class Build : NukeBuild
...
     readonly string Containers = ContainersLinux;
 
+    [Parameter("TargetFramework to be tested. Default is empty, meaning all TFMs supported by each test")]
+    readonly TargetFramework TestTargetFramework = TargetFramework.NOT_SPECIFIED;
+
     const string ContainersNone = "none";
     const string ContainersAny = "any";
     const string ContainersLinux = "linux";
     const string ContainersWindows = "windows";
+    const string ContainersWindowsContainerTestsOnly = "windows-container-tests-only";
```
Comment (pellared) [inline build/Build.cs:28]: nit: how about renaming to

```suggestion
    const string ContainersWindowsOnly = "windows-only";
```

File: build/Build.cs [hunk 9]
```diff
@@ -15,13 +15,17 @@ partial class Build : NukeBuild
     [Parameter("Platform to build - x86 or x64. Default is 'x64'")]
     readonly MSBuildTargetPlatform Platform = MSBuildTargetPlatform.x64;
 
-    [Parameter($"Docker containers type to be used. One of '{ContainersNone}', '{ContainersLinux}', '{ContainersWindows}'. Default is '{ContainersLinux}'")]
+    [Parameter($"Docker containers type to be used. One of '{ContainersNone}', '{ContainersLinux}', '{ContainersWindows}', '{ContainersWindowsContainerTestsOnly}'. Default is '{ContainersLinux}'")]
```
Comment (pellared) [inline build/Build.cs:18]:
```suggestion
    [Parameter($"Docker containers type to be used in tests. One of '{ContainersNone}', '{ContainersLinux}', '{ContainersWindows}', '{ContainersWindowsContainerTestsOnly}'. Default is '{ContainersLinux}'")]
```

**Review by pellared — APPROVED** [review]

**Review by pellared — APPROVED** [review]

File: build/Build.cs (hunk 8 above)
Comment (pjanotti) [reply build/Build.cs]: 8760e6f45

**Pull request merged**